atc/worker/artifact_source.go

```diff
@@ -73,13 +82,16 @@ func (source *artifactSource) StreamFile(
 	logger lager.Logger,
 	filepath string,
 ) (io.ReadCloser, error) {
-	out, err := source.volume.StreamOut(ctx, filepath)
+	out, err := source.volume.StreamOut(ctx, filepath, source.compression.Encoding())
 	if err != nil {
 		return nil, err
 	}
 
-	zstdReader := zstd.NewReader(out)
-	tarReader := tar.NewReader(zstdReader)
+	compressionReader, err := source.compression.NewReader(out)
+	if err != nil {
+		return nil, runtime.FileNotFoundError{Path: filepath}
```

Comment:
I'm not so sure if this is right :thinking:  

From what I understood, `NewReader` would error in multiple cases not really related to the file not having been found before (e.g., zstd failing to allocate buffers, or something like that).

On the other hand, `StreamOut` (few lines above), is something that is capable of returning a "file not found" type of error. e.g., consider the following call stack:

```
 0  0x0000000001a9be9b in github.com/concourse/baggageclaim/client.(*client).streamOut
    at /go/pkg/mod/github.com/concourse/baggageclaim@v1.6.5/client/client.go:288
 1  0x0000000001a9e6e1 in github.com/concourse/baggageclaim/client.(*clientVolume).StreamOut
    at /go/pkg/mod/github.com/concourse/baggageclaim@v1.6.5/client/volume.go:47
 2  0x0000000001adfa00 in github.com/concourse/concourse/atc/worker.(*volume).StreamOut
    at ./atc/worker/volume.go:92
 3  0x0000000001acbe82 in github.com/concourse/concourse/atc/worker.(*artifactSource).StreamFile
    at ./atc/worker/artifact_source.go:85
 4  0x0000000001ad3299 in github.com/concourse/concourse/atc/worker.(*client).StreamFileFromArtifact
    at ./atc/worker/client.go:709
 5  0x0000000001b9df3c in github.com/concourse/concourse/atc/exec.FileConfigSource.FetchConfig
    at ./atc/exec/task_config_source.go:84
 6  0x0000000001ba750b in github.com/concourse/concourse/atc/exec.(*FileConfigSource).FetchConfig
    at <autogenerated>:1
 7  0x0000000001b9e8ea in github.com/concourse/concourse/atc/exec.(*OverrideParamsConfigSource).FetchConfig
    at ./atc/exec/task_config_source.go:121
 8  0x0000000001b9f409 in github.com/concourse/concourse/atc/exec.InterpolateTemplateConfigSource.FetchConfig
    at ./atc/exec/task_config_source.go:168
 9  0x0000000001ba7862 in github.com/concourse/concourse/atc/exec.(*InterpolateTemplateConfigSource).FetchConfig
    at <autogenerated>:1
10  0x0000000001b9fdf9 in github.com/concourse/concourse/atc/exec.ValidatingConfigSource.FetchConfig
    at ./atc/exec/task_config_source.go:205
11  0x0000000001ba7ba9 in github.com/concourse/concourse/atc/exec.(*ValidatingConfigSource).FetchConfig
    at <autogenerated>:1
12  0x0000000001ba1799 in github.com/concourse/concourse/atc/exec.(*TaskStep).run
    at ./atc/exec/task_step.go:181
13  0x0000000001ba0dae in github.com/concourse/concourse/atc/exec.(*TaskStep).Run
```

looking at `github.com/concourse/baggageclaim/client.(*client).streamOut()`, we can see that the client does indeed perform some error checking:

```go
func getError(response *http.Response) error {
	var errorResponse *api.ErrorResponse
	err := json.NewDecoder(response.Body).Decode(&errorResponse)
	if err != nil {
		return err
	}

	if errorResponse.Message == api.ErrStreamOutNotFound.Error() {
		return baggageclaim.ErrFileNotFound
	}

	if response.StatusCode == 404 {
		return baggageclaim.ErrVolumeNotFound
	}

	return errors.New(errorResponse.Message)
}
```

https://github.com/concourse/baggageclaim/blob/aa42b1495315f94ff5ed8842df8b564d46cd572c/client/client.go#L298-L300

Reply:
hmm this line comes from the next couple lines that handles the tarReader.Next(). Given that there are multiple cases here that could not be captured by a single runtime error, we might just want to bubble up the error then?

Reply:
Since now we wrap around the reader from streamOut by `compression.NewReader()`, it is then decided by the underneath compression library that what error will be returned i.e. for now it is zstd.NewReader() and gzip.NewReader().

If the file is not found, the error would be captured on `compression.NewReader()` and it will never reach tar.NewReader() anymore.

I will remove the test and also remove the code to wrap error to be file not found err.

Reply:
:+1: 

@xtreme-sameer-vohra,  I think you were the one most involved in the refactor that introduced the `runtime.FileNotFoundError`, do you see any problem with this approach? Ultimately, the UX is the same (in the UI, you see `task config 'config/something-else.yaml' not found`), and given that there's no place where we branch based on `runtime.FileNotFoundError`, it seems good to me

Reply:
Removed the documentation of `runtime.FileNotFoundError`

Reply:
ping @xtreme-sameer-vohra

Reply:
Yes, agree with @cirocosta , the `out, err := source.volume.StreamOut(ctx, filepath, source.compression.Encoding())` would yield the FileNotFound error.

The [runtime.FileNotFoundError[(https://github.com/concourse/concourse/blame/master/atc/runtime/errors.go#L7) was introduced a long time ago and I can't think of a case where we need it for the `tarReader.Next` error.

Reply:
I haven't reviewed the entire PR, but this section looks good

Reply:
One option would be to move the encoding from the `stream out` call to the initialization of the baggageclaim client as the compression doesn't change once Concourse server is up. 

That would require bumping the baggageclaim library. However, since it doesn't change the contract between the client (baggageclaim) and the server (baggageclaim), it wouldn't require bumping worker version.